README.md
# Swift-AssetsLegalMentionsChecker

Swift-written tool to use in command line to check if images contain copyright / legal mentions / credits in metadata.
This project is built with Xcode.

## Managed files

This current version deals only with PNG file (with .png or .PNG extensions), and looks for some fields in the metadata : TIFF, IPTC and PNG.
If will check in these fields if the mention to look for is _exactly_ the defined value.

## Executables

If you are looking for the executables you can have a look on folder _~/Library/Developer/Xcode/DerivedData/Swift-AssetsLegalMentionsChecker-XXX/Build/Products/Debug/_ where XXX is a kind of random string generated by Xcode. You have to build the project with Xcode before.

## Run the program

Assuming _EXECUTABLE_ is the executable file (you can pick for example in the mentioned folder above):

To get the usage of the program:

```shell
EXECUTABLE --help
```

To get the version of the program:

```shell
EXECUTABLE --version
```

To run the program where it have to find recursively a CREDIT line in metadata of images in the FOLDER:

```shell
EXECUTABLE --folder FOLDER --mention CREDIT
```

You can also make the program displays more messages with the verbose option:

```shell
EXECUTABLE --folder FOLDER --mention CREDIT --verbose
```

## Exit codes

The program returns a value indicating if a problem occured or not:

* (-1): something wrong occured (command line options for example)
* (0): no error appeared, normal exit without image processing (e.g. display help)
* (+1): normal exit but failure occured during check of images (i.e. at least 1 file does not have legal mention in credits)
* (+2): normal exit, all files contain the legal mention

## Xcode integration

You can use this tool with an Xcode build script. The following sample can be placed in a script to add to your Xcode configuration. Thus the tool (with binary in the project) can be used easily.

```shell
#!/bin/sh
LEGAL_NOTICE="My Project, (c) Copyright Me-MySelf-And-I- SA 2020, CC-BY-SA-NC 4.0"
TOOL_PATH="path/to/binary/of/this/tool"
TARGET="."

$TOOL_PATH --folder "$TARGET" --mention "$LEGAL_NOTICE"
result=$?

if [ $result -eq 2 ]
then
	echo "✅ All assets contain legal notice in metadata."
	exit 0
else
	echo "🔴 Something wrong occured, see logs for further details."
	exit 1
fi
```


CHANGELOG.md
# What's new?

## Version 1.0.0

- Makes searches recursively in the target folder
- Looks for files filtering them using a regular expression for their extensions
- Deals only with PNG files
- Looks in TIFF, IPTC and PNG bundles of the metadata details for each file

## Version ?.?.?

- More details in documentation
- Add SwiftLint to project

Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/main.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

/// Program which has the aim of finding in images if metadata contain legal mentions line.
/// Indeed we might need to be sure all the assets included in a project have the legal mentions like
/// copyrights or licenses. These kinds of details are stored in the metadata of images.
///
/// For now only the PNG messages are managed. Such details are expected in metadata fields named TIFF, IPTC and PNG.
///
/// Program exists:
///     - (-1): something wrong occured (command line options for example)
///     - (0): no error appeared, normal exit without image processing (e.g. display help)
///     - (+1): normal exit but failure occured during check of images (i.e. at least 1 file does not have legal mention in credits)
///     - (+2): normal exit, all files contain the legam mention
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 19/06/2019
///

import Foundation

// MARK: - Configuration

public let VERSION = "1.0.0"
public var VERBOSE = false

private let consoleWritter = ConsoleOutput()
private let argumentsParser = ConsoleArgumentsParser()

// MARK: - Deal with options

consoleWritter.printWelcome()

let parameters = ConsoleInput().processProgram(arguments: &CommandLine.arguments)

guard argumentsParser.areWellDefined(arguments: parameters) else {
    consoleWritter.printBadCommandLineErrorMessage()
    exit(-1)
}

if argumentsParser.isForHelp(arguments: parameters) {
    consoleWritter.printUsage()
    consoleWritter.printBye()
    exit(0)
}

if argumentsParser.isForVersion(arguments: parameters) {
    consoleWritter.printVersion()
    consoleWritter.printBye()
    exit(0)
}

if argumentsParser.isVerboseDefined(in: parameters) {
    VERBOSE = true
}

let folderToProcess = parameters.filter { $0.0 == .folderToProcess }[0].1
let mentionToFind = parameters.filter { $0.0 == .lineToFind }[0].1

if folderToProcess.isEmpty {
    consoleWritter.write("The folder to process is undefined", to: .error)
    exit(-1)
}

if !FileManager.default.fileExists(atPath: folderToProcess) {
    consoleWritter.write("The folder to process does not exist, please check its path", to: .error)
    exit(-1)
}

if mentionToFind.isEmpty {
    consoleWritter.write("The credit line to look for is undefined", to: .error)
    exit(-1)
}

// MARK: - Instructions to check files

consoleWritter.write("Will look in folder '\(folderToProcess)' for mention '\(mentionToFind)'", to: .standard)

let areAllResourcesSuitable = AssetsLegalMentionsChecker("png|PNG").lookIn(folder: folderToProcess, for: mentionToFind)

// MARK: - Check of results

if !areAllResourcesSuitable {
    consoleWritter.write("🚨 FAILURE 🚨: There is at least one file without legal mention in metadata")
    consoleWritter.printBye()
    exit(1)
} else {
    consoleWritter.write("✅ SUCCESS ✅: All files contain legal mention in metadata")
    consoleWritter.printBye()
    exit(2)
}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Console/ConsoleIO.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// Structure to use so as to write in outpu channels some messages
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 19/06/2019
///
struct ConsoleOutput {

    /// Displays a welcome message
    ///
    func printWelcome() {
        write("*************************************************************************")
        write("* Swift-AssetsLegalMentionsChecker                                      *")
        write("* Copyright (C) 2019 Orange SA                                          *")
        write("* MIT License                                                           *")
        write("* Version: \(VERSION)                                                        *")
        write("* Find easily in your images assets if legal mentions have been defined *")
        write("*************************************************************************")
    }

    /// Displays the exit common message
    ///
    func printBye() {
        write("Execution completed, bye!")
    }

    /// Displays in the standard output the usage of the program
    ///
    func printUsage() {
        let executableName = (CommandLine.arguments[0] as NSString).lastPathComponent
        write("Usage:")
        write("\t\(executableName) --folder path --mention line")
        write("\t\t - path: The path to the root folder containing the assets to process")
        write("\t\t - mention: The legal mention line to lok in metadata of images on the fodler at path")
        write("or")
        write("\t\(executableName) --help")
        write("\t\t To display this help message")
        write("or")
        write("\t\(executableName) --version")
        write("\t\t To display the version of the program")
    }

    /// Displays in the standard output the version of the program
    ///
    func printVersion() {
        write("Version: \(VERSION)")
    }

    /// Prints error message with more details
    ///
    func printBadCommandLineErrorMessage() {
        write("Command line not written as expected", to: .error)
        printUsage()
        printBye()
    }

    /// Writes in the selected output the message
    /// - Parameters:
    ///     - message: The message to write
    ///     - to: The output to use
    ///
    func write(_ message: String, to: OutputTypes = .standard) {
        switch to {
        case .standard:
            print("\(message)")
        case .error:
            fputs("💥 Error: \(message)\n", stderr)
        }
    }

    /// Writes the message in the channel if the VERBOSE flag is enabled
    /// - Parameters:
    ///     - message: The message to write
    ///     - to: The output to use
    ///
    func verbose(_ message: String, to: OutputTypes = .standard) {
        if VERBOSE {
            write(message, to: to)
        }
    }

}

/// Structure to use so as to read console entries or arguments
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 20/06/2019
///
struct ConsoleInput {

    /// Process the bundle of arguments given to the program and check their values.
    /// Arguments must be a 5-length array or a 2-length array.
    /// Item 0 of this array is the program name, other elements are the parameters.
    ///
    /// - Parameters:
    ///     - arguments: The bunch of arguments of the program
    /// - Returns:
    ///     - An array of tuples containing the argument type and its value
    ///
    func processProgram(arguments args: inout [String]) -> [(ConsoleArgumentTypes, String)] {

        let argsCount = args.count

        if argsCount == 2 && args[1] == "--help" {
            return [(ConsoleArgumentTypes.help, "")]
        }

        if argsCount == 2 && args[1] == "--version" {
            return [(ConsoleArgumentTypes.version, "")]
        }

        let expectedMinimalArgumentsCount = 5
        if argsCount != expectedMinimalArgumentsCount
            && argsCount != (expectedMinimalArgumentsCount + 1) { // + 1 -> maybe add of --verbose
            return [(ConsoleArgumentTypes.undefined, "")]
        }

        var options: [(ConsoleArgumentTypes, String)] = []
        if  isVerboseDefined(in: args) {
            options.append((.verbose, ""))
            args = args.filter { $0 != "--verbose" }
        }

        for index in stride(from: 1, to: argsCount-1, by: 2) {
            let option = ConsoleArgumentTypes(value: args[index])
            guard option != .undefined else {
                return [(option, "")]
            }
            if option != .verbose {
                options.append((option, args[index+1]))
            }
        }

        return options

    }

    /// Checks if the verbose option has been added oe not
    /// - Returns:
    ///     - a boolean value, true if available false otherwise
    ///
    private func isVerboseDefined(in arguments: [String]) -> Bool {
        return  arguments.filter { $0 == "--verbose" }.count == 1
    }

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Console/OutputTypes.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
     
     MIT License
     
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// The types of outputs dealt by the logger features.
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 19/06/2019
///
enum OutputTypes {

    /// Should use the error output
    case error

    /// Should use the standard output
    case standard

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Checker/FolderCrawler.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// Structure which has the aim of crawling in a folder to look for image ressources for further treatments.
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.2
/// - Since: 27/06/2019
///
struct FolderCrawler {

    /// Goes recursively in the folder in parameter and keep only ressources matching the condition
    /// - Parameters:
    ///     - folder: The folder to process, path can be absolute or relative
    ///     - filtering: A regular expression to apply to the extension of the files (e.g. png|jpg)
    /// - Returns:
    ///     - An array of pathes to ressources matching the conditions
    ///
    func crawlIn(folder path: String, filtering byExtension: String) -> [String] {

        let filesEnumerator = FileManager.default.enumerator(atPath: path)
        let filesPaths = filesEnumerator?.allObjects as! [String]
        let ressourcesPaths = filesPaths.filter {
            let fileUrl = NSURL(fileURLWithPath: $0)
            guard let fileExtension = fileUrl.pathExtension else {
                return false
            }
            return fileExtension.range(of: byExtension, options: .regularExpression, range: nil, locale: nil) != nil
        }

        let consoleOutput = ConsoleOutput()
        consoleOutput.write("Founds \(ressourcesPaths.count) files with the extension '\(byExtension)'")

        var results: [String] = []
        for resourcePath in ressourcesPaths {
            let globalPath = "\(path)/\(resourcePath)" // FIXME: Seems dirty
            results.append(globalPath)
            consoleOutput.verbose("Found file located at '\(globalPath)'")
        }

        return results

    }

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Checker/MetadataExplorer.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// Structure which has the aim of reading metadata of images to look for expected details
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 27/06/2019
///
struct MetadataExplorer {

    // MARK: - Variables

    /// To display messages in the console
    let output = ConsoleOutput()

    // MARK: - Functions

    /// Reads the metadata of the all the files at these paths and looks if an expected detail is available or not.
    /// - Parameters:
    ///     - at: The array of files' paths which must contain the expected detail in metadata
    ///     - for: The detail expected in the metadata of the file at this path
    /// - Returns:
    ///     - A boolean value indicating if the detail is available in all the files(true) or not (false for at least 1 file)
    ///
    func look(at files: [String], for detail: String) -> Bool {
        output.verbose("Will process \(files.count) files")
        var allFilesAreSuitable = true
        for file in files {
            output.verbose("Processing file '\(file)")
            let isFileSuitable = look(at: file, for: detail)
            if !isFileSuitable {
                output.write("❌ It seems the file at \(file) does not have in metadata the mention")
            }
            allFilesAreSuitable = allFilesAreSuitable && isFileSuitable
        }
        return allFilesAreSuitable
    }

    /// Reads the metadata of the file at this path and looks if an expected detail is available or not.
    /// Will look in dedicated metadata bundles: TIFF, IPTC and PNG. These groups are defined for PNG files.
    /// - Parameters:
    ///     - at: The path to the image file (must be PNG) which must contain the expected detail in metadata
    ///     - for: The detail expected in the metadata of the file at this path
    /// - Returns:
    ///     - A boolean value indicating if the detail is available (true) or not (false)
    ///
    func look(at path: String, for detail: String) -> Bool {

        let url = URL(fileURLWithPath: path)

        guard let imageData = try? Data(contentsOf: url) else {
            output.write("A problem occured with the file at '\(path)'", to: .error)
            return false
        }
        guard let source = CGImageSourceCreateWithData(imageData as CFData, nil) else {
            output.write("A problem occured with the file at '\(path)'", to: .error)
            return false
        }
        guard let metadata = CGImageSourceCopyPropertiesAtIndex(source, 0, nil)
            as? [AnyHashable: Any] else {
            output.write("It seems no metadata are avaialble for file at '\(path)'", to: .error)
            return false
        }

        var areDetailsDefined = false
        if processTIFF(bundle: metadata["{TIFF}"] as? [String: String]) == detail {
            areDetailsDefined = true
        } else {
            output.verbose("⚠️  Warning: Legal mentions are not defined in TIFF metadata field")
        }
        if processIPTC(bundle: metadata["{IPTC}"] as? [String: String]) == detail {
            areDetailsDefined = true
        } else {
            output.verbose("⚠️  Warning: Legal mentions are not defined in IPTC metadata field")
        }
        if processPNG(bundle: metadata["{PNG}"] as? [String: Any]) == detail {
            areDetailsDefined = true
        } else {
            output.verbose("⚠️  Warning: legal mentions are not defined in PNG metadata field")
        }

        return areDetailsDefined

    }

    /// Parses the group of metadata assuming the group is reated to TIFF field.
    /// Returns the value of the field which should contain the legal mentions
    /// - Parameters:
    ///     - data: The bunch of data to process
    /// - Returns:
    ///     - The value of the field which should contain the legal mention, or nil if undefined
    ///
    private func processTIFF(bundle data: [String: String]?) -> String? {
        guard let data = data else {
            return nil
        }
        return data["Copyright"]
    }

    /// Parses the group of metadata assuming the group is reated to IPTC field.
    /// Returns the value of the field which should contain the legal mentions
    /// - Parameters:
    ///     - data: The bunch of data to process
    /// - Returns:
    ///     - The value of the field which should contain the legal mention, or nil if undefined
    ///
    private func processIPTC(bundle data: [String: String]?) -> String? {
        guard let data = data else {
            return nil
        }
        return data["CopyrightNotice"]
    }

    /// Parses the group of metadata assuming the group is reated to PNG field.
    /// Returns the value of the field which should contain the legal mentions
    /// - Parameters:
    ///     - data: The bunch of data to process
    /// - Returns:
    ///     - The value of the field which should contain the legal mention, or nil if undefined
    ///
    private func processPNG(bundle data: [String: Any]?) -> String? {
        guard let data = data else {
            return nil
        }
        return data["Copyright"] as? String
    }

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/AssetsLegalMentionsChecker.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// Structure which runs the instructions to read target folder, filter images and check if they
/// contain legal mentions in  metadata.
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 28/06/2019
///
struct AssetsLegalMentionsChecker {

    // MARK: - Variables

    /// The filter to use for the files to check
    private let ressourcesFilter: String


    // MARK: - Cnostructor

    /// Initializes the regular expression to use so as to filter the files extensions
    /// and keep only these we want to check
    /// - Parameters:
    ///     - filter: The regular expression to apply, e.g "png|PNG"
    ///
    init(_ filter: String) {
        ressourcesFilter = filter
    }

    // MARK: - Methods

    /// Looks recursively in the target folder for images  files with extension
    /// matching the filter "png|PNG", and check for these files if they contain in
    /// metadata the legal mentions.
    /// - Parameters:
    ///     - folder: The target folder to start looking
    ///     - mention - The credit line to find in metadata data
    /// - Returns:
    ///     - A flag indicating if at least one file does not contain the credit (false) or if
    /// all the files are suitable (true)
    ///
    func lookIn(folder target: String, for credit: String) -> Bool {
        let matchingRessources = FolderCrawler().crawlIn(folder: target, filtering: ressourcesFilter)
        let areAllResourcesSuitable = MetadataExplorer().look(at: matchingRessources, for: credit)
        return areAllResourcesSuitable
    }

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Console/ConsoleArgumentsParser.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// Structure to use so as to parse raw parameters given to the command line in console mode
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 28/06/2019
///
struct ConsoleArgumentsParser {

    /// Checks if the arguments given in aprameters of the command are well written and can be used, or not
    /// - Returns: A flag indicating of parameters canot be used (false) or can be (true)
    ///
    func areWellDefined(arguments params: [(ConsoleArgumentTypes, String)]) -> Bool {
        return
            (params.count == 1 && params[0].0 == .help)
            ||
            (params.count == 1 && params[0].0 == .version)
            ||
            (params.count == 2 && (
                (params[0].0 == .folderToProcess && params[1].0 == .lineToFind) ||
                    (params[0].0 == .lineToFind && params[1].0 == .folderToProcess)
            ))
            ||
            (params.count == 3 && isVerboseDefined(in: params)) // FIXME Not efficient, missing cases
    }

    /// Returns true if the option is for help message to display or false otherwise
    /// - Returns: A boolean value
    ///
    func isForHelp(arguments params: [(ConsoleArgumentTypes, String)]) -> Bool {
        return (params.count == 1 && params[0].0 == .help)
    }

    /// Returns true if the option is for version message to display or false otherwise
    /// - Returns: A boolean value
    ///
    func isForVersion(arguments params: [(ConsoleArgumentTypes, String)]) -> Bool {
        return (params.count == 1 && params[0].0 == .version)
    }

    /// Returns true if there is an option for the verbose mode or not (false)
    /// - Returns: A boolean value
    ///
    func isVerboseDefined(in arguments: [(ConsoleArgumentTypes, String)]) -> Bool {
        return arguments.filter { $0.0 == .verbose }.count == 1
    }

}


Swift-AssetsLegalMentionsChecker/Swift-AssetsLegalMentionsChecker/Console/ConsoleArgumentTypes.swift
/**
     Swift-AssetsLegalMentionsChecker
     Copyright (C) 2019 Orange SA
 
     MIT License
 
     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:
     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.
     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
 */

import Foundation

/// The types of options to give to the program.
///
/// - Author: Pierre-Yves Lapersonne
/// - Version: 1.0.1
/// - Since: 19/06/2019
///
enum ConsoleArgumentTypes: String {

    /// The location of the folder to process recursively
    case folderToProcess = "--folder"

    /// The credit / copyright / legal mention to look in metadata of images in the folder
    case lineToFind = "--mention"

    /// Display the help / usage
    case help = "--help"

    /// Display the version
    case version = "--version"

    /// Define the verbose mode
    case verbose = "--versbose"

    /// Undefined option
    case undefined

    /// Constructor
    /// - Parameters:
    ///     - value: The value given to the program as parameter
    ///
    init(value: String) {
        switch value {
        case "--folder":
            self = .folderToProcess
        case "--mention":
            self = .lineToFind
        case "--help":
            self = .help
        case "--version":
            self = .version
        case "--verbose":
            self = .verbose
        default:
            self = .undefined
        }
    }

}
